lib/matplotlib/_constrained_layout.py

```diff
@@ -106,42 +108,132 @@ def do_constrained_layout(fig, renderer, h_pad, w_pad,
 
         # make margins for all the axes and subfigures in the
         # figure.  Add margins for colorbars...
-        _make_layout_margins(fig, renderer, h_pad=h_pad, w_pad=w_pad,
-                             hspace=hspace, wspace=wspace)
-        _make_margin_suptitles(fig, renderer, h_pad=h_pad, w_pad=w_pad)
+        make_layout_margins(layoutgrids, fig, renderer, h_pad=h_pad,
+                            w_pad=w_pad, hspace=hspace, wspace=wspace)
+        make_margin_suptitles(layoutgrids, fig, renderer, h_pad=h_pad,
+                              w_pad=w_pad)
 
         # if a layout is such that a columns (or rows) margin has no
         # constraints, we need to make all such instances in the grid
         # match in margin size.
-        _match_submerged_margins(fig)
+        match_submerged_margins(layoutgrids, fig)
 
         # update all the variables in the layout.
-        fig._layoutgrid.update_variables()
+        layoutgrids[fig].update_variables()
 
-        if _check_no_collapsed_axes(fig):
-            _reposition_axes(fig, renderer, h_pad=h_pad, w_pad=w_pad,
-                             hspace=hspace, wspace=wspace)
+        if check_no_collapsed_axes(layoutgrids, fig):
+            reposition_axes(layoutgrids, fig, renderer, h_pad=h_pad,
+                            w_pad=w_pad, hspace=hspace, wspace=wspace)
         else:
             _api.warn_external('constrained_layout not applied because '
                                'axes sizes collapsed to zero.  Try making '
                                'figure larger or axes decorations smaller.')
-        _reset_margins(fig)
+        reset_margins(layoutgrids, fig)
+    return layoutgrids
+
 
+def make_layoutgrids(fig, layoutgrids):
```

Comment:
All the others take `layoutgrids` first; should this not do the same also?

Reply:
This is (usually) called first, so the `layoutgrids` is only if we are doing the call recursively....